promql/engine.go

```diff
@@ -0,0 +1,1241 @@
+	if res.Err != nil {
+		return res.Err.Error()
+	}
+	if res.Value == nil {
+		return ""
+	}
+	return res.Value.String()
+}
+
+type (
+	// ErrQueryTimeout is returned if a query timed out during processing.
+	ErrQueryTimeout string
+	// ErrQueryCanceled is returned if a query was canceled during processing.
+	ErrQueryCanceled string
+)
+
+func (e ErrQueryTimeout) Error() string  { return fmt.Sprintf("query timed out in %s", e) }
+func (e ErrQueryCanceled) Error() string { return fmt.Sprintf("query was canceled in %s", e) }
+
+// A Query is derived from an a raw query string and can be run against an engine
+// it is associated with.
+type Query interface {
+	// Exec starts evaluating the query and returns immediately.
+	// The returned channel blocks until the evaluation is done and receives
+	// false if an error occurred, true otherwise.
+	Exec() (done <-chan bool)
+	// Result returns the result of the evaluated query. It blocks until the
+	// result is available.
+	Result() Result
+	// Statements returns the parsed statements of the query.
+	Statements() Statements
+	// Stats returns statistics about the lifetime of the query.
+	Stats() *stats.TimerGroup
+	// Cancel signals that a running query execution should be aborted.
+	Cancel()
+}
+
+// query implements the Query interface.
+type query struct {
+	q      string     // The original query string.
+	stmts  Statements // Statements of the parsed query.
+	result Result     // Result of the evaluated query.
+	done   chan bool
+	stats  *stats.TimerGroup
+	cancel func()
+
+	ng *Engine // The engine against which the query is executed.
+}
+
+// Statements implements the Query interface.
+func (q *query) Statements() Statements {
+	return q.stmts
+}
+
+// Stats implements the Query interface.
+func (q *query) Stats() *stats.TimerGroup {
+	return q.stats
+}
+
+// Cancel implements the Query interface.
+func (q *query) Cancel() {
+	if q.cancel != nil {
+		q.cancel()
+	}
+}
+
+// Exec implements the Query interface.
+func (q *query) Exec() <-chan bool {
+	ctx := q.ng.setupQuery(q)
+	go func() {
+		q.result.Value, q.result.Err = q.ng.exec(ctx, q)
+		success := q.result.Err == nil
+		q.done <- success
+		q.done <- success
+	}()
+	return q.done
+}
+
+// Result retrieves the result from an executed query. It blocks as long as
+// the execution is running.
+func (q *query) Result() Result {
+	if q.done == nil {
+		return Result{Err: errors.New("query execution not started")}
+	}
+	<-q.done
+	return q.result
+}
+
+type (
+	// AlertHandlers can be registered with an engine and are called on
+	// each executed alert statement.
+	AlertHandler func(context.Context, *AlertStmt) error
+	// RecordHandlers can be registered with an engine and are called on
+	// each executed record statement.
+	RecordHandler func(context.Context, *RecordStmt) error
+)
+
+// ContextDone returns an error if the context was canceled or timed out.
+func ContextDone(ctx context.Context, env string) error {
```

Comment:
Is this method really meant to be public, or only used internally to determine whether the query should abort?